src/lib.rs

```diff
@@ -0,0 +1,1175 @@
+        }
+        v.len = len;
+        v.mask_last_block();
+        v
+    }
+}
+
+impl<T: Debug + PrimInt + One + Zero> Vob<T> {
+    /// Constructs a new, empty Vob (with a user-defined backing storage type) with the given
+    /// capacity.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::<u32>::new_with_storage_type(0);
+    /// v.push(true);
+    /// assert_eq!(v[0], true);
+    /// ```
+    pub fn new_with_storage_type(capacity: usize) -> Vob<T> {
+        Vob {
+            len: 0,
+            vec: Vec::with_capacity(capacity)
+        }
+    }
+
+    /// Returns the number of bits the Vob can hold without reallocating.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// assert_eq!(Vob::new().capacity(), 0);
+    /// assert!(Vob::with_capacity(1).capacity() >= 1);
+    /// ```
+    pub fn capacity(&self) -> usize {
+        // This multiplication can't overflow because of the checks in reserve()
+        self.vec.capacity() * bits_per_block::<T>()
+    }
+
+    /// Reserves capacity for at least `additional` more bits to be inserted in the Vob. The Vob
+    /// may reserve more space to avoid frequent reallocations. After calling `reserve`, capacity
+    /// will be greater than or equal to `self.len() + additional`. Does nothing if capacity is
+    /// already sufficient.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::new();
+    /// v.reserve(1);
+    /// assert!(v.capacity() >= 1);
+    /// ```
+    pub fn reserve(&mut self, additional: usize) {
+        // Repeatedly calling this will over-reserve, since we only record how many bits we've
+        // used, not how many bits of capacity the user thinks we have. For example, repeatedly
+        // asking for storage for one extra bit will repeatedly add a whole extra machine word.
+        // Fixing this would require an extra machine word in the Vob struct, which hardly seems
+        // worth it for an unlikely, and probably silly, use case.
+        let cap = self.vec.capacity()
+                          .checked_mul(bits_per_block::<T>())
+                          .and_then(|x| x.checked_add(additional))
+                          .expect("Overflow detected");
+        self.vec.reserve(blocks_required::<T>(cap));
+    }
+
+    /// Shrinks the capacity of the vector as much as possible.
+    ///
+    /// It will drop down as close as possible to the length but the allocator may still inform the
+    /// vector that there is space for a few more elements.
+    pub fn shrink_to_fit(&mut self) {
+        self.vec.shrink_to_fit();
+    }
+
+    /// Shortens the Vob, keeping the first `len` elements and dropping the rest.
+    ///
+    /// If len is greater than the vector's current length, this has no effect.
+    ///
+    /// The drain method can emulate truncate, but causes the excess elements to be returned instead of dropped.
```

Comment:
Long line? Vim seems to want to wrap it around 100.

Reply:
Oops, quite right!